source/Coop/Mod/GameSync/Hideout/HideoutBehavior.cs

```diff
@@ -0,0 +1,87 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using TaleWorlds.CampaignSystem;
+using TaleWorlds.Core;
+using TaleWorlds.ObjectSystem;
+
+namespace Coop.Mod.GameSync.Hideout
+{
+    class HideoutBehavior : CampaignBehaviorBase
+    {
+        /// <summary>
+        ///     Store that lets you know which player has already seen a hideout.
+        /// </summary>
+        private readonly Dictionary<MBGUID, List<MBGUID>> settlementsSpotted = new Dictionary<MBGUID, List<MBGUID>>();
+
+        /// <summary>
+        ///     Register events related to Hideouts.
+        /// </summary>
+        public override void RegisterEvents()
+        {
+            CampaignEvents.HourlyTickSettlementEvent.AddNonSerializedListener(this, OnHourlyTickSettlement);
+            CampaignEvents.OnHideoutDeactivatedEvent.AddNonSerializedListener(this, OnHideoutDeactivated);
+        }
+
+        /// <summary>
+        ///     Listener for <see cref="CampaignEvents.HourlyTickSettlementEvent"/>.
+        ///     Used for detecting an hideout around players.
+        /// </summary>
+        /// <param name="settlement">Settlement</param>
+        public void OnHourlyTickSettlement(Settlement settlement)
+        {
+            if (!Coop.IsServer)
+            {
+                return;
+            }
+
+            if (settlement.IsHideout && settlement.Hideout.IsInfested && !settlement.Hideout.IsSpotted)
+            {
+                float hideoutSpottingDistance = Campaign.Current.Models.MapVisibilityModel.GetHideoutSpottingDistance();
+                
+                foreach (var playerControlledParty in CoopServer.Instance.Persistence.MobilePartyEntityManager.PlayerControlledParties)
+                {
+                    float partyDistanceSquared = playerControlledParty.Position2D.DistanceSquared(settlement.Position2D);
+                    float num2 = 1f - partyDistanceSquared / (hideoutSpottingDistance * hideoutSpottingDistance);
```

Comment:
That seems like a very complicated way of writing `bool isSpotted = partyDistanceSquared < hideoutSpottingDistance * hideoutSpottingDistance`.